### ISSUE TITLE
Common SDVariable/INDArray interface

### ISSUE THREAD

**Issue by rnett** [eclipse/deeplearning4j#7971]
Proposal: create a common interface for `SDVariable` and `INDArray`.
This would allow easy switching between the two (for example, define a neural network layer that can be used for INDArrays and SDVariable).

Contains tensor ops that are callable on both, e.g. `eq`, `reshape`.

To implement this (without a lot of unsupported ops on one vs. the other) we'll need a base interface.
For now, call this ITensor (final name is TBD):
* INDArray extends ITensor - with methods like getDouble/putScalar, toDoubleVector(), toFlatArray() etc
* SDVariable extends ITensor - with methods like getArr(), eval() etc

Signature: `ITensor<T extends ITensor<T, INDEX>, INDEX> `, `INDEX` is necessary for `get`.

Potential issues:
Add some transforms that aren't in either, but (imo) should be: #7970 
Methods that are dynamic for SDVariable but static for INDArray (e.g. `SDVariable.shape()` is a `SDVariable` but `INDArray.shape()` is a `long[]`

**Comment by AlexDBlack** [comment]
cc @treo @raver119

**Comment by raver119** [comment]
From my point of view: there's a Graph. It consists of Nodes. Nodes take Inputs and generate Outputs. Both Inputs and Outputs are Variables. Technically - yes, Variables can be INDArrays. But they also can be: Lists, Maps, Queues, abstract Arrow things. Are we also going to add methods related to these entities to SDVariable?

**Comment by AlexDBlack** [comment]
So the main use case here was being able to write code once, that works for INDArray or SDVariable inputs/output.
You can imagine a layer implemented once, that you can call/execute either way, instead of having 2 functionaly equivalent implementations.

That's quite a reasonable goal at first glance - there is a lot of overlap, in terms of API/ops, which is what prompted this idea in the first place.

Suppose we wanted to implement a DL4J+SameDiff layer in this manner - it couldn't be done in practice using only the generic/parent interface:
- Can't use .size(int), shape(), rank() methods that return int/long, which virtually every DL4J layer uses (often for handling different input cases, such as in batch norm, or looping, as in RNNs)
- "inplace" ops are not supported on SDVariable (partly because we need them for backprop, partly we'll auto-optimize this during execution eventually); array reuse (for performance) also can't be used
- We can only use methods on INDArray/SDVariable - no Nd4j.X/SameDiff.X methods, no Transforms methods, no ND4J.exec(Op) etc
- For things like RNNs (that use java loops), we'd essentially being unrolling them - making them work only for a fixed sequence length for the SDVariable case
- workspaces can't be used (only applies to INDArray not SDVariable)
- DL4J Helpers (like cudnn/mkl-dnn) obviously apply to INDArrays only (we need to get pointers etc to execute them)
- In the case of a "generic INDArray/SDVaribale" DL4J layer, you still need to define backprop using INDArrays anyway (to be actually useful/trainable when executing using INDArrays).

So on reflection, this only seems viable for the simplest use cases and layers only. It's a nice idea with an admirable goal (improved ux and reducing code duplication), but it's not workable in practice. Consequently I'll close this, and we can remove it from the roadmap.

Only question remaining for me is whether there's anything we can salvage of this proposal...
The idea of having a mirrored API for both INDArray and SDVariable may have some value, simply to minimize switching costs for users who know one API but not yet the other. We've already been doing that informally anyway, enforcing this via a common interface (*an interface that nobody is supposed to actually use in practice*) is probably not be worth it.
